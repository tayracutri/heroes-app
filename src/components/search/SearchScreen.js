import { useMemo } from 'react';
import queryString from 'query-string'
import { useLocation } from 'react-router';
import { useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import HeroesCard from '../heroes/HeroesCard';
import { getHeroByName } from '../../selectors/getHeroByName';

export const SearchScreen = ({ history }) => {
    const navigate = useNavigate()
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const filteredHeroes = useMemo(() => getHeroByName(q), [q])

    const [values, handleInputChange] = useForm({
        formText: q
    });
    const { formText } = values;

    const handleSubmit = (e) => {
        e.preventDefault();
        navigate(`?q=${formText}`)
    }

    return (
        <div>
            <h1 className='m-4'>Search Screen</h1>

            <div className='row'>
                <div className='col-4'>


                    <form onSubmit={handleSubmit} className='mt-4'>
                        <input
                            type='text'
                            placeholder='Insert a superhero'
                            className='form-control'
                            name='formText'
                            autoComplete='off'
                            value={formText}
                            onChange={handleInputChange}
                        ></input>
                        <button type='submit' className='btn m-1 btn-block btn-outline-primary btn-block' > Submit</button>
                    </form>

                </div>
                <div className='col-8'>
                    <h5 className='mt-4'>Results</h5>
                    <hr />

                    {

                        (q === '') //empty string, search a hero
                            ? <div className='alert alert-info'> Search a hero</div>
                            //if the user searched an unexistent hero (length = 0), throw couldnt find it
                            : (filteredHeroes.length === 0)
                            && <div className='alert alert-danger'> Couldnt find a hero named {q}</div>
                    }
                    {
                        filteredHeroes.map(hero => (
                            <HeroesCard
                                key={hero.id}
                                {...hero} />
                        ))
                    }

                </div>
            </div>
        </div>
    )
}
