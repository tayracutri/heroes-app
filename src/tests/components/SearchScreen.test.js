import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../components/search/SearchScreen";

const mockNavigate = jest.fn()

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockNavigate
}))

describe('test on SerachScreen', () => {
    test('should show the components w the default values', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero')
    });

    test('should show Batman and the input w the queryString valie', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('should show an error if the hero isnt found', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmn']}>
                <SearchScreen />
            </MemoryRouter>
        )
        expect(wrapper.find('.alert-danger').text().trim()).toBe('Couldnt find a hero named batmn')
    })

    test('should call the navigate to the new screen', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>

                <SearchScreen />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {
            target: {
                name: 'formText',
                value: 'batman'
            }
        });

        wrapper.find('form').simulate('submit')
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');

    });


});
