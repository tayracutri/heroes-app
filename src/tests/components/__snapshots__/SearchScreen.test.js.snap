// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on SerachScreen should show Batman and the input w the queryString valie 1`] = `
<div>
  <h1
    className="m-4"
  >
    Search Screen
  </h1>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <form
        className="mt-4"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="formText"
          onChange={[Function]}
          placeholder="Insert a superhero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary btn-block"
          type="submit"
        >
           Submit
        </button>
      </form>
    </div>
    <div
      className="col-8"
    >
      <h5
        className="mt-4"
      >
        Results
      </h5>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="../assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test on SerachScreen should show the components w the default values 1`] = `
<div>
  <h1
    className="m-4"
  >
    Search Screen
  </h1>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <form
        className="mt-4"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="formText"
          onChange={[Function]}
          placeholder="Insert a superhero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary btn-block"
          type="submit"
        >
           Submit
        </button>
      </form>
    </div>
    <div
      className="col-8"
    >
      <h5
        className="mt-4"
      >
        Results
      </h5>
      <hr />
      <div
        className="alert alert-info"
      >
         Search a hero
      </div>
    </div>
  </div>
</div>
`;
